pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,154 @@
+// short-name mode as configured in the registries.conf:
+//
+// * "" or "disabled": append the name to all specified unqualified-search
+//                     registries (USR).
+//
+// * "enforcing": prompt the user to select a USR.  Error if stdout is not a
+//                TTY as we cannot prompt.
+//
+// * "permissive": prompt the user to select a USR.  If stdout is not a TTY,
+//                 fall back to the "disabled" behaviour.
+//
+// Note that if there is only one USR, the USR is auto-selected without prompt.
+// If there is no USR, "docker.io" is selected.
+//
+// If `clear` is set, a possibly existing alias will be cleared.
+func Resolve(ctx *types.SystemContext, name string, clear bool) ([]string, error) {
+	// If `input` is not a short name, we're done.
+	isShort, err := IsShortName(name)
+	if err != nil {
+		return nil, err
+	}
+	if !isShort {
+		return []string{name}, nil
+	}
+
+	// If there's already an alias, use it.
+	isAliased, alias, err := sysregistriesv2.ResolveShortNameAlias(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+	if isAliased {
+		if !clear {
+			return []string{alias}, nil
+		}
+		if err := sysregistriesv2.ClearShortNameAlias(ctx, name); err != nil {
+			return nil, errors.Wrapf(err, "error clearing short-name alias %q", name)
+		}
+	}
+
+	// No alias, so we have to consult the search registries.
+	unqualifiedSearchRegistries, err := sysregistriesv2.UnqualifiedSearchRegistries(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	candidates := []string{}
+	switch len(unqualifiedSearchRegistries) {
+	case 0:
+		// No search registries, so pick the default one.
+		return []string{completeNameWithRegistry(name, defaultRegistry)}, nil
+	case 1:
+		// Only one, so we really don't have a choice.
+		return []string{completeNameWithRegistry(name, unqualifiedSearchRegistries[0])}, nil
+	default:
+		// Create a slice of candidates the user may want to refer to.
+		for _, reg := range unqualifiedSearchRegistries {
+			candidates = append(candidates, completeNameWithRegistry(name, reg))
+		}
+	}
+
+	mode, err := sysregistriesv2.GetShortNameMode(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we're not prompting, just return all possible
+	// combinations.
+	if mode == sysregistriesv2.ShortNameModeDisabled {
+		return candidates, nil
+	}
+
+	if !terminal.IsTerminal(int(os.Stdout.Fd())) {
+		if mode == sysregistriesv2.ShortNameModePermissive {
+			return candidates, nil
+		}
+		if mode == sysregistriesv2.ShortNameModeEnforcing {
+			return nil, errors.New("short-name resolution enforced but cannot prompt without a TTY")
+		}
+	}
+
+	// A simple prompt with all possible options.
+	prompt := promptui.Select{
+		Label:    "Please select an image",
+		Items:    candidates,
+		HideHelp: true, // do not show navigation help
+	}
+
+	_, selection, err := prompt.Run()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := sysregistriesv2.AddShortNameAlias(ctx, name, selection); err != nil {
```

Comment:
So the UI model is that the user must choose a registry to pull from, that choice is recorded, and _then_ we try pulling. If the user chose “wrong” and the image does not actually exist there, it’s going to be annoying to recover.

(Yes, I know, this is clearly more secure than letting the user guess and then record an alias for the first image that actually successfully pulls. Still, is it intentional/desired?)

---

Another aspect to this: In the “disabled” setting, I assume the full set of options is used to pull. Wouldn’t we still want to record the outcome of the pull after the search completes, so that future uses of the same short name always use the same registry? (I’m sure this was covered in the design doc. I need to look it up, link it to this PR, and then revisit.)

Reply:
> Still, is it intentional/desired?

Intentional, yes. Desired, up to debate I guess.

For now it makes testing and development easier. Recording the name/alias pair and adding them _after_ a successful pull would be a more intrusive change, and I didn't want to invest time into something that I believe is an open question.

I'll set up a meeting next week where we can go through the details :+1: 

> In the “disabled” setting, I assume the full set of options is used to pull. Wouldn’t we still want to record the outcome of the pull after the search completes, so that future uses of the same short name always use the same registry? 

I intended "disabled" to fully opt-out from aliasing. I will push the latest development state in a jiffy where the three modes should make more sense. If unspecified, we now default to "permissive" which I see as a compromise of aliasing "where possible". Taking the above point a bit further in the permissive context, it may really be better to only add an alias once it's know to be pulled successfully. This way, "permissive" would record aliases even without a TTY/prompt.